src/hotspot/share/utilities/istream.hpp

```diff
@@ -0,0 +1,447 @@
+
+  // Reset indexes within the buffer to point to no content.
+  void clear_buffer();
+
+  // Reset indexes within the buffer to point to the given content.
+  // This is where we scan for newlines as well.
+  void set_buffer_content(size_t content_start, size_t content_end);
+
+  // Try to make the buffer bigger.  This may be necessary in order to
+  // buffer a very long line.  Returns false if there was an
+  // allocation failure.
+  //
+  // On allocation failure, just make do with whatever buffer there
+  // was to start with; the caller must check for this condition and
+  // avoid buffering more data in the non-expanded buffer.  However,
+  // the buffer will always be non-null, so at least one line can be
+  // buffered, if it is of normal size.
+  bool expand_buffer(size_t new_length);
+
+  // Make sure there is at least one line in the buffer, and set
+  // _beg/_end to indicate where it is.  Any content before _beg can
+  // be overwritten to make more room in the buffer.  If there is no
+  // more input, set the state up to indicate we are done.
+  bool fill_buffer();
+
+  // Find some room in the buffer so we call read on it.
+  // This might call expand_buffer but will try not to.
+  // The assumption is that read already buffers slow I/O calls.
+  // The purpose for the small buffer managed here is to store whole lines,
+  // and perhaps edit them in-place.
+  void prepare_to_fill_buffer(size_t& fill_offset, size_t& fill_length);
+
+  // Quick check for an initially incomplete buffer...
+  void preload() const {
+    if (need_to_read()) {
+      const_cast<inputStream*>(this)->fill_buffer();
+    }
+  }
+
+  // How much content is buffered (if any) after the current line?
+  size_t buffered_content_length(bool include_current) const {
+    return (include_current       ? _content_end - _beg :
+            _content_end >= _next ? _content_end - _next : 0);
+  }
+
+  // Returns a pointer and count to characters buffered after the
+  // current line, but not yet read from my input source.  Only useful
+  // if you are trying to stack input streams on top of each other
+  // somehow.  You can also ask the input source if it thinks it has
+  // more bytes.
+  const char* next_content(size_t& next_content_length) const;
+
+ public:
+  // Create an empty input stream.
+  // Call push_back_input or set_input to configure.
+  inputStream() {
+    _input = nullptr;
+    _buffer = nullptr;
+    _must_free = nullptr;
+    _buffer_size = 0;
+    _line_count = 0;
+    _beg = _end = _next = _content_end = 0;
+    _line_ending = 0;
+    _input_state = NTR_STATE;
+  }
+
+  // Take input from the given source.  Buffer only a modest amount.
+  inputStream(Input* input)
+    : inputStream()
+  {
+    set_input(input);
+  }
+
+  virtual ~inputStream() {
+    if (_must_free)         handle_free();
+    if (_input != nullptr)  set_input(nullptr);
+  }
+
+  // Discards any previous input and sets the given input source.
+  void set_input(Input* input);
```

Comment:
There may be a better way to design the way streams and inputs connect with each other, but removing `set_input` is a step in the wrong direction.  The reason is that a stream is designed to be stack-allocated, while an input source is something that can have a non-stack lifetime.

It is supposed to be cheap to create an input stream on stack, and part of the cost model comes from the small buffer, which is also stack allocated, and defers the need for heap allocation.  This means it could be a performance problem to “make a new one” as suggested above.  It’s better to allow the stack allocated one to modify its input source, if the input is either (a) multiplexed from different sources, or (b) determined AFTER the input stream is declared.

Yes, the normal use is to determine the input source first and then wrap the i-stream around it.  But if this fails, it is not the right answer to push the i-stream onto the heap; i-streams are strongly associated with the blocks of code that exercise them.  The right answer, when the input sources are “hard to manage”, is `set_input`.